core/server/master/src/main/java/alluxio/master/file/meta/InvalidationSyncCache.java

```diff
@@ -0,0 +1,413 @@
+    }
+  }
+
+  /**
+   * There is a most one writer at a time to a sync state.
+   * But multiple readers may be reading the values while they
+   * are being updated, thus the values are volatile.
+   */
+  static class SyncState {
+    volatile long mDirectChildInvalidation = 0;
+    volatile long mRecursiveChildInvalidation = 0;
+    volatile long mInvalidationTime = 0;
+    volatile long mValidationTime = 0;
+    volatile long mDirectValidationTime = 0;
+    volatile long mRecursiveValidationTime = 0;
+    volatile boolean mIsFile;
+
+    SyncState(boolean isFile) {
+      mIsFile = isFile;
+    }
+
+    void setInvalidationTime(long time) {
+      if (time > mInvalidationTime) {
+        mInvalidationTime = time;
+      }
+    }
+
+    void setDirectChildInvalidation(long time) {
+      if (time > mDirectChildInvalidation) {
+        mDirectChildInvalidation = time;
+      }
+    }
+
+    void setRecursiveChildInvalidation(long time) {
+      if (time > mRecursiveChildInvalidation) {
+        mRecursiveChildInvalidation = time;
+      }
+    }
+
+    void setIsFile(boolean isFile) {
+      // only transition from file to directory
+      // and not the other way around
+      if (!isFile && mIsFile) {
+        mIsFile = false;
+      }
+    }
+
+    void setValidationTime(long time, DescendantType descendantType) {
+      if (time > mValidationTime) {
+        mValidationTime = time;
+      }
+      if (descendantType == DescendantType.ALL) {
+        if (time > mRecursiveValidationTime) {
+          mRecursiveValidationTime = time;
+        }
+        if (time > mDirectValidationTime) {
+          mDirectValidationTime = time;
+        }
+      } else if (descendantType == DescendantType.ONE) {
+        if (time > mDirectValidationTime) {
+          mDirectValidationTime = time;
+        }
+      }
+    }
+  }
+
+  /**
+   * Called when starting a sync.
+   * @return the time at the start of the sync
+   */
+  public long startSync() {
+    return mClock.millis();
+  }
+
+  /**
+   * Get sync times for a given path if they exist in the cache.
+   * @param path the path to check
+   * @return a pair of sync times, where element 0 is the direct sync time
+   * and element 1 is the recursive sync time
+   */
+  public Optional<Pair<Long, Long>> getSyncTimesForPath(AlluxioURI path) {
+    SyncState state = path.getPath().equals(AlluxioURI.SEPARATOR)
+        ? mRoot : mItems.getIfPresent(path.getPath());
+    return Optional.ofNullable(state).map(syncState ->
+        new Pair<>(syncState.mValidationTime, syncState.mRecursiveValidationTime));
+  }
+
+  /**
+   * Check if sync should happen.
+   * A path is checked starting from the full path, all the way up to the root.
+   * At each path the largest validation time and invalidation time is computed depending
+   * on its descendant type. At the end if the invalidation is more recent than the validation
+   * then a sync is needed. Otherwise, a sync is needed based on the difference between
```

Comment:
At the end if the invalidation is more recent than the validation then a sync is needed.
-> did you miss something for this sentence?